pkg/volume/scaleio/sio_plugin.go

```diff
@@ -0,0 +1,186 @@
+	return nil
+}
+
+// *******************
+// VolumePlugin Impl
+// *******************
+var _ volume.VolumePlugin = &sioPlugin{}
+
+func (p *sioPlugin) Init(host volume.VolumeHost) error {
+	p.host = host
+	p.mounter = host.GetMounter()
+	p.volumeMtx = keymutex.NewKeyMutex()
+	return nil
+}
+
+func (p *sioPlugin) GetPluginName() string {
+	return sioPluginName
+}
+
+func (p *sioPlugin) GetVolumeName(spec *volume.Spec) (string, error) {
+	source, err := getVolumeSourceFromSpec(spec)
+	if err != nil {
+		return "", err
+	}
+
+	return source.VolumeName, nil
+}
+
+func (p *sioPlugin) CanSupport(spec *volume.Spec) bool {
+	return (spec.PersistentVolume != nil && spec.PersistentVolume.Spec.ScaleIO != nil) ||
+		(spec.Volume != nil && spec.Volume.ScaleIO != nil)
+}
+
+func (p *sioPlugin) RequiresRemount() bool {
+	return false
+}
+
+func (p *sioPlugin) NewMounter(
+	spec *volume.Spec,
+	pod *api.Pod,
+	_ volume.VolumeOptions) (volume.Mounter, error) {
+	sioSource, err := getVolumeSourceFromSpec(spec)
+	if err != nil {
+		glog.Error(log("failed to extract ScaleIOVolumeSource from spec: %v", err))
+		return nil, err
+	}
+
+	return &sioVolume{
+		pod:         pod,
+		spec:        spec,
+		source:      sioSource,
+		namespace:   pod.Namespace,
+		volSpecName: spec.Name(),
+		volName:     sioSource.VolumeName,
+		podUID:      pod.UID,
+		readOnly:    sioSource.ReadOnly,
+		fsType:      sioSource.FSType,
+		plugin:      p,
+	}, nil
+}
+
+// NewUnmounter creates a representation of the volume to unmount
+// The specName param can be used to carry the namespace value (if needed) using format:
+// specName = [<namespace>nsSep]<somevalue> where the specname is pre-pended with the namespace
+func (p *sioPlugin) NewUnmounter(specName string, podUID types.UID) (volume.Unmounter, error) {
+	glog.V(4).Info(log("Unmounter for %s", specName))
+
+	return &sioVolume{
+		podUID:      podUID,
+		volSpecName: specName,
+		plugin:      p,
+	}, nil
+}
+
+func (p *sioPlugin) ConstructVolumeSpec(volumeName, mountPath string) (*volume.Spec, error) {
+	sioVol := &api.Volume{
+		Name: volumeName,
+		VolumeSource: api.VolumeSource{
+			ScaleIO: &api.ScaleIOVolumeSource{},
+		},
+	}
+	return volume.NewSpecFromVolume(sioVol), nil
+}
+
+//******************************
+// PersistentVolumePlugin Impl
+// *****************************
+var _ volume.PersistentVolumePlugin = &sioPlugin{}
+
+func (p *sioPlugin) GetAccessModes() []api.PersistentVolumeAccessMode {
+	return []api.PersistentVolumeAccessMode{
+		api.ReadWriteOnce,
+	}
+}
+
+// ***************************
+// DeletableVolumePlugin Impl
+//****************************
+var _ volume.DeletableVolumePlugin = &sioPlugin{}
+
+func (p *sioPlugin) NewDeleter(spec *volume.Spec) (volume.Deleter, error) {
+	sioSource, err := getVolumeSourceFromSpec(spec)
+	if err != nil {
+		glog.Error(log("deleter failed to extract source from spec: %v", err))
+		return nil, err
+	}
+
+	specName := spec.Name()
+	namespace, _ := splitVolumeName(specName)
```

Comment:
Making an arbitrary namespace in pv name doesn't guarantee access to that namespace.

You can get namespace from `spec.ClaimRef.Namespace` to access PVC's namespace.